chore(batch): update Batch README.md (#6931)
aws/aws-cdk: packages/@aws-cdk/aws-batch/README.md

---
<!--END STABILITY BANNER-->

> This module is part of the [AWS Cloud Development Kit](https://github.com/aws/aws-cdk) project.

Author: @iliapolo I will add this back in directly underneath the banner (if thats ok). Otherwise, we can rearrange it.

> Batch achieves this by utilizing queue processing of batch job requests. To successfully handle a job request, you need the following resources:

Author: Definitely better wording! Love it <3

> 1. [Job Definition](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-batch.JobDefinition.html) - *the container image to handle a job request*

Author: > This is hard to phrase since it's not just the container image. I imagine you singled it out since its the only one that cannot be overridden during job submission?

That was exactly the reasoning. Lets bring this change in then.

> 2. [Job Queue](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-batch.JobQueue.html) - *the queue where batch jobs can be submitted to via AWS SDK/CLI*

Author: Alright 👍

> 3. [Compute Environment](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-batch.ComputeEnvironment.**html**) - *the cluster where provisioned compute resources will exist*

Author: Alright, good idea. Best not to over-complicate the understanding 👌

> 

Author: Sure, I will adjust it as such. 👍

For more information on **AWS Batch** visit the [AWS Docs for Batch](https://docs.aws.amazon.com/batch/index.html).

## Compute Environment

> At the core of AWS Batch is the compute environment. All batch jobs are processed within a compute resource like EC2 or Spot. These resources are managed inside an ECS cluster. In **MANAGED** mode, AWS will handle the communications with AWS EC2/Spot for provisioning compute resources based on demand. Otherwise, **UNMANAGED** mode requires you to manage the allocation of compute resources for your related ECS cluster.

Author: True, lets make this adjustment.

Below is an example of each available type of compute environment:

```ts
const defaultVpc = new ec2.Vpc(this, 'VPC');

> 

Author: Thank you for double-checking <3 I will change this.

```
Compute Environment:

1. m5.xlarge => 4 vCPU
2. m5.2xlarge => 8 vCPU
```

```
Job Queue:
---------
| A | B |
---------

Job Requirements:
A => 4 vCPU - ALLOCATED TO m5.xlarge
B => 2 vCPU - WAITING
```

In this situation, Batch will allocate **Job A** to compute resource #1 because it is the most cost efficient resource that matches the vCPU requirement. However, with this `BEST_FIT` strategy, **Job B** will not be allocated to our other available compute resource even though it is strong enough to handle it. Instead, it will wait until the first job is finished processing or wait a similar `m5.xlarge` resource to be provisioned.

The alternative would be to use the `BEST_FIT_PROGRESSIVE` strategy in order for the remaining job to be handled in larger containers regardless of vCPU requirement and costs.

### Advanced Compute Environment Configuration

> By default, CDK will use the the `optimal` instance type family to choose compute resource types that best match your job requirements. However, you can configure what instance types to provision, and the structure of those instances when assigned to your compute environm$$ent. Specifying an advanced compute environment could look like this:

Author: Sorry, I had money on my mind 🤪

> By default, CDK will use the the `optimal` instance type family to choose compute resource types that best match your job requirements. However, you can configure what instance types to provision, and the structure of those instances when assigned to your compute environm$$ent. Specifying an advanced compute environment could look like this:

Author: I can see this in the AWS Console UI when setting up a brand new managed compute environment. `optimal` was the default shown.

> By default, CDK will use the the `optimal` instance type family to choose compute resource types that best match your job requirements. However, you can configure what instance types to provision, and the structure of those instances when assigned to your compute environm$$ent. Specifying an advanced compute environment could look like this:

Author: So the initial attempt was to mimic the functionality of the UI.

```ts

> const keyPair = new ec2.Key

Author: Definitely missing code. Removing this now...

    image: new ecs.EcsOptimizedAmi({
      generation: ec2.AmazonLinuxGeneration.AMAZON_LINUX_2,
      hardwareType: ecs.AmiHardwareType.GPU,
    }),

    // 'ec2KeyPair' specifies the name of the related EC2 KeyPair for SSH access
    ec2KeyPair: 'dev',

    // 'instanceRole' defines the privileges the job will have when running
    instanceRole: new iam.Role(stack, 'compute-env-instance-role', {
      assumedBy: new iam.ServicePrincipal('batch.amazonaws.com'),
    }),

    // Only pick `p3.2xlarge` or `p3.8xlarge` instances to provision
    instanceTypes: [
      ec2.InstanceType.of(ec2.InstanceClass.P3, ec2.InstanceSize.XLARGE2),
      ec2.InstanceType.of(ec2.InstanceClass.P3, ec2.InstanceSize.XLARGE8),
    ],

    // At most, provision up to 40 vCPUs worth of instances. For example, 5x p3.8xlarge => 40 vCPU
    maxvCpus: 40,

    // 'minvCpus' specifies our required minimal vCPU availability when no jobs exists
    minvCpus: 8,

    // 'vpc' and 'vpcSubnets' locks down our containers to the private subnet of our VPC
    vpc,
    vpcSubnets: {
      subnetType: ec2.SubnetType.PRIVATE,
    },
  },
});
```

### Launch template support

> #### Usage

Author: Ok 👍

Simply define your Launch Template:
```typescript
const myLaunchTemplate = new ec2.CfnLaunchTemplate(this, 'LaunchTemplate', {
  launchTemplateName: 'extra-storage-template',
  launchTemplateData: {
    blockDeviceMappings: [
      {
        deviceName: '/dev/xvdcz',
        ebs: {
          encrypted: true,
          volumeSize: 100,
          volumeType: 'gp2'
        }
      }
    ]
  }
});
```

and use it:

```typescript
const myComputeEnv = new batch.ComputeEnvironment(this, 'ComputeEnv', {
  computeResources: {
    launchTemplate: {
      launchTemplateName: myLaunchTemplate.launchTemplateName as string, //or simply use an existing template name
    },
    vpc,
  },
  computeEnvironmentName: 'MyStorageCapableComputeEnvironment',
});
```

> ### Handling unexpected peaks using Spot-Based resources

Author: Based on the comments, I removed this second to avoid confusion. It was more of a scenario building section to help others understanding. But I see best that we shouldn't presume.

```ts
const myLaunchTemplate = new ec2.CfnLaunchTemplate(this, 'LaunchTemplate', {
  managed: true,
  computeResources: {
    type: batch.ComputeResourceType.SPOT,
    // Bid at 75% of the on-demand price
    bidPercentage: 75,
    // Limits the minimum/maximum size of our spotfleet will be
    maxvCpus: 4,
    minvCpus: 1,
    vpc,
  },
});
```

### Importing an existing Compute Environment

To import an existing batch compute environment, call `ComputeEnvironment.fromComputeEnvironmentArn()`.

Below is an example:

```ts
const computeEnv = ComputeEnvironment.fromComputeEnvironmentArn(this, 'imported-compute-env', 'arn:aws:batch:us-east-1:555555555555:compute-environment/My-Compute-Env');
```

## Job Queue

Batch Job Queues store incoming requests before assignment to an available compute environment.

```ts
const jobQueue = new batch.JobQueue(stack, 'JobQueue', {
  computeEnvironments: [
    {
      // Defines a collection of compute resources to handle assigned batch jobs
      computeEnvironment,
      // Order determines the allocation order for jobs (i.e. Lower means higher preferance for job assignment)
      order: 1,
    },
  ],
});
```

### Priorty-Based Queue Example

> Sometimes you might come across a situation where certain batch jobs need to be processed before others. In this case, creating priority based queues can achieve this. Below is an example of how that can be done:

Author: Thank you. I agree, we should stress the importance here that this is about precedence and not ordered processing.

```ts
const highPrioQueue = new batch.JobQueue(stack, 'JobQueue', {
  computeEnvironments: sharedComputeEnvs,
  priority: 2,
});

const lowPrioQueue = new batch.JobQueue(stack, 'JobQueue', {
  computeEnvironments: sharedComputeEnvs,
  priority: 1,
});
```

> In this example, both job queues share the same set of compute environments but our `highPrioQueue` will have its jobs assigned first. Even though our other queue might have jobs waiting to be assigned, they will be ignored until our `highPrioQueue` has no jobs waiting for assignment.

Author: That is actually true... queues do use multiple environments if linked 🤔 . I will go ahead and modify/rework this then.

### Importing an existing Job Queue

To import an existing batch job queue, call `JobQueue.fromJobQueueArn()`.

Below is an example:

```ts
const jobQueue = JobQueue.fromJobQueueArn(this, 'imported-job-queue', 'arn:aws:batch:us-east-1:555555555555:job-queue/High-Prio-Queue');
```

## Job Definition

A Batch Job definition helps AWS Batch understand important details about how to run your application in the scope of a Batch Job. This involves key information like resource requirements, what containers to run, how the compute environment should be prepared, and more. Below is a simple example of how to create a job definition:

```ts

> const repo = new ecr.Repository(stack, 'batch-job-repo');

Reviewer: Maybe use an existing repo as an example? It more clearly shows a link to a specific image:

```suggestion
const repo = ecr.Repository.fromRepositoryName(stack, 'batch-job-repo', 'todo-list');
```

new batch.JobDefinition(stack, 'batch-job-def-from-ecr', {
  container: {
    image: new ecs.EcrImage(repo, 'latest'),
  },
});
```

### Using a local Docker project

> Below is an example of how you can create a Batch Job Definition from a local Docker projects on deploy of your CDK project:

Reviewer: ```suggestion
Below is an example of how you can create a Batch Job Definition from a local Docker application.
```
